vowpalwabbit/core/src/reductions/interaction_ground.cc

```diff
@@ -3,110 +3,369 @@
+
+    float pa = action_scores[i].score;
+
+    if (i == chosen_action_idx) {
+      igl.ik_ex.weight = 1/4.f / pa;
+    } else {
+      igl.ik_ex.weight = 3/4.f / (1 - pa);
+    }
+
+    igl.ik_ex.interactions = &ik_interactions;
+    igl.ik_ex.extent_interactions = igl.extent_interactions; // TODO(low pri): not reuse igl.extent_interactions, need to add in feedback
+
+    // 2. ik learn
+    igl.ik_learner->learn(igl.ik_ex, 0);
+
+    if (i == chosen_action_idx) {
+      ik_pred = igl.ik_ex.pred.scalar;
     }
+
+    action_ex->l.cb = action_ex->l.cb_with_observations.event;
+    action_ex->l.cb_with_observations.event.reset_to_default();
+  }
+
+  std::swap(igl.ik_ftrl->all->loss, igl.ik_all->loss);
+  std::swap(igl.ik_ftrl->all->sd, igl.ik_all->sd);
+
+  float predicted_cost = 0.f;
+  // 4. update multi line ex label
+  if (ik_pred * 2 > 1.f) {
+    bool is_definitely_bad = observation_ex->l.cb_with_observations.is_definitely_bad;
+    predicted_cost = -1.f + is_definitely_bad * (1.f + 1.f / p_unlabeled_prior);
+  }
+
+  // 5. Train pi policy
+  ec_seq[chosen_action_idx]->l.cb.costs[0].cost = predicted_cost;
+
+  std::swap(*igl.pi_ftrl.get(), *igl.ik_ftrl);
+
+  // TODO: switch the label parser type to CB for text format
+
+  // reset the first_observed label for cb_explore_adf reduction
+  (*igl.ik_ftrl).all->sd->first_observed_label = FLT_MAX;
+  base.learn(ec_seq, 1);
+  igl.known_cost = VW::get_observed_cost_or_default_cb_adf(ec_seq);
+
+  std::swap(*igl.pi_ftrl.get(), *igl.ik_ftrl);
+
+  for (auto& ex : ec_seq) {
+    ex->l.cb_with_observations.event = ex->l.cb;
+    ex->l.cb.reset_to_default();
   }
+
+  ec_seq.push_back(observation_ex);
 }
 
-void learn(interaction_ground& ig, learner& base, VW::multi_ex& ec_seq)
+void predict(interaction_ground& igl, learner& base, VW::multi_ex& ec_seq)
 {
-  // find reward of sequence
-  VW::cb_class label = VW::get_observed_cost_or_default_cb_adf(ec_seq);
-  ig.total_uniform_cost += label.cost / label.probability / ec_seq.size();  //=p(uniform) * IPS estimate
-  ig.total_uniform_reward += -label.cost / label.probability / ec_seq.size();
-
-  // find prediction & update for cost
-  base.predict(ec_seq);
-  ig.total_importance_weighted_cost += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
-  base.learn(ec_seq);
-
-  // find prediction & update for reward
-  label.cost = -label.cost;
+  VW::example* observation_ex = nullptr;
+
+  if (ec_seq.size() > 0 && ec_seq.back()->l.cb_with_observations.is_observation) {
+    observation_ex = ec_seq.back();
+    ec_seq.pop_back();
+  }
+
+  std::swap(*igl.pi_ftrl.get(), *igl.ik_ftrl);
+
+  for (auto& ex : ec_seq) {
+    ex->l.cb = ex->l.cb_with_observations.event;
+    ex->l.cb_with_observations.event.reset_to_default();
+  }
+
   base.predict(ec_seq, 1);
-  ig.total_importance_weighted_reward += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
+  std::swap(*igl.pi_ftrl.get(), *igl.ik_ftrl);
```

Comment:
does this swap work? I'm not sure why we implemented our own swap in the first place but if so, we can remove the swap declaration from interaction_ground.h

Reply:
yes, removed swap_ftrl